java/algorithm/src/main/java/org/apache/arrow/algorithm/dictionary/SearchTreeBasedDictionaryBuilder.java

```diff
@@ -0,0 +1,157 @@
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.arrow.algorithm.dictionary;
+
+import java.util.TreeSet;
+
+import org.apache.arrow.algorithm.sort.VectorValueComparator;
+import org.apache.arrow.vector.ValueVector;
+
+/**
+ * A dictionary builder is intended for the scenario frequently encountered in practice:
+ * the dictionary is not known a priori, so it is generated dynamically.
+ * In particular, when a new value arrives, it is tested to check if it is already
+ * in the dictionary. If so, it is simply neglected, otherwise, it is added to the dictionary.
+ *
+ * <p>
+ *   This class builds the dictionary based on a binary search tree.
+ *   Each add operation can be finished in O(log(n)) time,
+ *   where n is the current dictionary size.
+ * </p>
+ * @param <V> the dictionary vector type.
+ */
+public class SearchTreeBasedDictionaryBuilder<V extends ValueVector> {
+
+  /**
+   * The dictionary to be built.
+   */
+  protected final V dictionary;
+
+  /**
+   * The criteria for sorting in the search tree.
+   */
+  protected final VectorValueComparator<V> comparator;
+
+  /**
+   * The total number of data items probed.
+   */
+  protected int totalValueCount;
+
+  /**
+   * The number of items in the dictionary.
+   */
+  protected int distinctValueCount;
+
+  /**
+   * If null should be encoded.
+   */
+  protected final boolean encodeNull;
+
+  /**
+   * The search tree for storing the value index.
+   */
+  protected TreeSet<Integer> searchTree;
+
+  /**
+   * Construct a search tree-based dictionary builder.
+   * @param dictionary the dictionary vector.
+   * @param comparator the criteria for value equality.
+   */
+  public SearchTreeBasedDictionaryBuilder(V dictionary, VectorValueComparator<V> comparator) {
+    this(dictionary, comparator, false);
+  }
+
+  /**
+   * Construct a search tree-based dictionary builder.
+   * @param dictionary the dictionary vector.
+   * @param comparator the criteria for value equality.
+   * @param encodeNull if null should be encoded.
+   */
+  public SearchTreeBasedDictionaryBuilder(V dictionary, VectorValueComparator<V> comparator, boolean encodeNull) {
+    this.dictionary = dictionary;
+    this.comparator = comparator;
+    this.encodeNull = encodeNull;
+    this.comparator.attachVector(dictionary);
+
+    searchTree = new TreeSet<>((index1, index2) -> comparator.compare(index1, index2));
+  }
+
+  public void startBuild() {
+    totalValueCount = 0;
+    distinctValueCount = 0;
+  }
+
+  public void endBuild() {
+    dictionary.setValueCount(distinctValueCount);
+  }
+
+  public V getDictionary() {
+    return dictionary;
```

Comment:
you might want a guard here to ensure this  and getSortedDictionary aren't called until endBuild is called.

Reply:
Good suggestion. We use a bool flag to mark the end of dictionary building. Please take a look.